### PR TITLE
 Functional aws_cloudtrail_trail resource

### LINKED ISSUE #146: Add resource for cloudtrail trail
For CIS, a singular cloudtrail trail resource is needed.  Need properties s3_bucket_name and log_group_name.

Refs https://github.com/chef/inspec/issues/2233

### PR DESCRIPTION
**Pull request by rx294**
Fixes #146

### CODE REVIEW

File: docs/resources/aws_cloudtrail_trail.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+The test will pass if the identified trail is a multi region trail.
+
+    describe aws_cloudtrail_trail('trail-name') do
+      it { should be_multi_region_trail }
+    end
+
+### be_encrypted
+
+The test will pass if the logs delivered by identified trail is encrypted.
```
Comment (clintoncwolfe) [inline docs/resources/aws_cloudtrail_trail.md]: delivered by *the* identified trail

File: docs/resources/aws_cloudtrail_trail.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    describe aws_cloudtrail_trail('trail-name') do
+      its('s3_bucket_name') { should cmp "s3-bucket-name" }
+    end
+
+### trail_arn
+
+The ARN identifier of the specified trail. An ARN uniquely identifies the trail within AWS.
+
+    describe aws_cloudtrail_trail('trail-name') do
+      its('trail_arn') { should cmp "trail-arn" }
```
Comment (clintoncwolfe) [inline docs/resources/aws_cloudtrail_trail.md]: I'd prefer that we had a real-looking ARN URI as the comparison value

File: docs/resources/aws_cloudtrail_trail.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+The test will pass if the logs delivered by identified trail is encrypted.
+
+    describe aws_cloudtrail_trail('trail-name') do
+      it { should be_encrypted }
+    end
+
+### be_log_file_validation_enabled
+
+The test will pass if the identified trail log file integrity validation is enabled.
```
Comment (clintoncwolfe) [inline docs/resources/aws_cloudtrail_trail.md]: The test will pass if the identified trail *has* log file integrity validation is enabled.

File: libraries/aws_cloudtrail_trail.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
+class AwsCloudTrailTrail < Inspec.resource(1)
+  name 'aws_cloudtrail_trail'
+  desc 'Verifies settings for individual AWS CloudTrail Trail'
```
Comment (clintoncwolfe) [inline libraries/aws_cloudtrail_trail.rb]: Verifies settings for *an* individual AWS CloudTrail Trail

File: test/integration/default/build/aws.tf [hunk 5]
```diff
@@ -7,6 +7,15 @@ provider "aws" {
...
   value = "${data.aws_caller_identity.creds.account_id}"
 }
+
+data "aws_region" "region" {
+  current = true
+}
+
+output "aws_region" {
+  value = "${data.aws_region.region.name}"
+}
```
Comment (clintoncwolfe) [inline test/integration/default/build/aws.tf]: 👍  to this, I've almost added it myself several times.

File: test/integration/default/build/cloudtrail.tf [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+      "Resource": "*",
+      "Condition": {
+        "StringEquals": {
+          "kms:ViaService": "ec2.us-east-1.amazonaws.com",
+          "kms:CallerAccount": "${data.aws_caller_identity.creds.account_id}"
+        }
+      }
+    },
+    {
+      "Sid": "Enable cross account log decryption",
```
Comment (clintoncwolfe) [inline test/integration/default/build/cloudtrail.tf]: Do we require this ability?

File: test/integration/default/verify/controls/aws_cloudtrail_trail.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  describe aws_cloudtrail_trail(fixtures['cloudtrail_trail_2_name']) do
+    it { should_not be_multi_region_trail }
+  end
+  describe aws_cloudtrail_trail(fixtures['cloudtrail_trail_1_name']) do
+    it { should be_log_file_validation_enabled }
+  end
+  describe aws_cloudtrail_trail(fixtures['cloudtrail_trail_2_name']) do
+    it { should_not be_log_file_validation_enabled }
+  end
+end
```
Comment (clintoncwolfe) [inline test/integration/default/verify/controls/aws_cloudtrail_trail.rb]: :100: The fixtures and integration tests are very good here.

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
Just a few tiny doc edits, and a question about an IAM policy.  Very good work, and a very good contribution - thank you!

File: test/integration/default/build/cloudtrail.tf (hunk 6 above)
Comment (rx294) [reply test/integration/default/build/cloudtrail.tf]: Not necessary...i was going with the default Key Policy Created in CloudTrail Console.
I have removed it.

**Comment by rx294** [comment]
Thank you @clintoncwolfe i have made the updates.

**Comment by clintoncwolfe** [comment]
I'm happy with the code, and unit tests pass cleanly.  However, the integration tests fail for me, with an IAM Role error:

```
aws_cloudtrail.trail_1: Still creating... (10s elapsed)
Error applying plan:

1 error(s) occurred:

* aws_cloudtrail.trail_1: 1 error(s) occurred:

* aws_cloudtrail.trail_1: InvalidCloudWatchLogsLogGroupArnException: Access denied. Check the permissions for your role.
	status code: 400, request id: 6263e786-d490-4a99-9ae9-90ae11da9df8
```

Offhand, that looks like [this now-fixed terraform-aws bug](https://github.com/terraform-providers/terraform-provider-aws/issues/693), but that had a workaround of re-running terraform, which doesn't work here; that makes me think the policy is actually bad, rather than being a timing issue.

I'll continue making my "rounds"; and then circle back to this later, if you haven't gotten to it yet.  IAM policies are always pretty subtle, and I'm no expert on what CloudTrail needs.

**Comment by rx294** [comment]
@clintoncwolfe thats weird the integration tests are working fine on my side.

File: test/integration/default/build/cloudtrail.tf [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Sid": "AWSCloudTrailCreateLogStream",
+            "Effect": "Allow",
+            "Action": [
+                "logs:CreateLogStream"
+            ],
+            "Resource": [
+                "arn:aws:logs:us-east-1:${data.aws_caller_identity.creds.account_id}:log-group:${aws_cloudwatch_log_group.trail_1_log_group.name}:log-stream:${data.aws_caller_identity.creds.account_id}_CloudTrail_${data.aws_region.region.name}*"
```
Comment (clintoncwolfe) [inline test/integration/default/build/cloudtrail.tf]: Here's the location triggering the IAM role permission issue - the `us-east-1` here needs to be parameterized, that is, replaced with `${data.aws_region.region.name}`.

I test in us-east-2, and this grants permission to work in us-east-1; thus the conflict.

File: test/integration/default/build/cloudtrail.tf [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+            ]
+        },
+        {
+            "Sid": "AWSCloudTrailPutLogEvents",
+            "Effect": "Allow",
+            "Action": [
+                "logs:PutLogEvents"
+            ],
+            "Resource": [
+                "arn:aws:logs:us-east-1:${data.aws_caller_identity.creds.account_id}:log-group:${aws_cloudwatch_log_group.trail_1_log_group.name}:log-stream:${data.aws_caller_identity.creds.account_id}_CloudTrail_${data.aws_region.region.name}*"
```
Comment (clintoncwolfe) [inline test/integration/default/build/cloudtrail.tf]: Same issue here, parameterize the region.

**Comment by clintoncwolfe** [comment]
Ok, I found where it was happening - a hardcoded region in the policy.  The integration tests pass once that's in place.  Once you make the change, we'll get it merged!

**Comment by rx294** [comment]
Ah! my bad...I have made the corrections...thanks again @clintoncwolfe.

**Review by clintoncwolfe — APPROVED** [review]

**Comment by clintoncwolfe** [comment]
Thanks for this excellent contribution!

**Pull request merged**